blockchain/blockchain.go

```diff
@@ -0,0 +1,275 @@
+package blockchain
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"path"
+	"strings"
+	"sync"
+
+	logging "github.com/ipfs/go-log/v2"
+	gostream "github.com/libp2p/go-libp2p-gostream"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+const (
+	FxBlockchainProtocolID = "/fx.land/blockchain/0.0.1"
+	FxBlockchainEndPoint   = "127.0.0.1:4000"
+
+	actionSeeded = "seeded"
+	actionAuth   = "auth"
+)
+
+var (
+	_ Blockchain = (*FxBlockchain)(nil)
+
+	log            = logging.Logger("fula/blockchain")
+	errInvalidType = errors.New("invalid type for to")
+)
+
+type (
+	FxBlockchain struct {
+		*options
+		h host.Host
+		s *http.Server
+		c *http.Client
+
+		authorizedPeers     map[peer.ID]struct{}
+		authorizedPeersLock sync.RWMutex
+	}
+	authorizationRequest struct {
+		Subject peer.ID `json:"id"`
+		Allow   bool    `json:"allow"`
+	}
+)
+
+func NewFxBlockchain(h host.Host, o ...Option) (*FxBlockchain, error) {
+	opts, err := newOptions(o...)
+	if err != nil {
+		return nil, err
+	}
+	bl := &FxBlockchain{
+		options: opts,
+		h:       h,
+		s:       &http.Server{},
+		c: &http.Client{
+			Transport: &http.Transport{
+				DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
+					pid, err := peer.Decode(strings.TrimSuffix(addr, ".invalid:80"))
+					if err != nil {
+						return nil, err
+					}
+					return gostream.Dial(ctx, h, pid, FxBlockchainProtocolID)
+				},
+			},
+		},
+		authorizedPeers: make(map[peer.ID]struct{}),
+	}
+	if bl.authorizer != "" {
+		if err := bl.SetAuth(context.Background(), h.ID(), bl.authorizer, true); err != nil {
+			return nil, err
+		}
+	}
+	return bl, nil
+}
+
+func (bl *FxBlockchain) Start(ctx context.Context) error {
+	listen, err := gostream.Listen(bl.h, FxBlockchainProtocolID)
+	if err != nil {
+		return err
+	}
+	bl.s.Handler = http.HandlerFunc(bl.serve)
+	go func() { bl.s.Serve(listen) }()
+	return nil
+}
+
+func (bl *FxBlockchain) Seeded(ctx context.Context, to interface{}, r SeededRequest) ([]byte, error) {
+	var method string
+	var addr string
+
+	switch v := to.(type) {
```

Comment:
Can you help me understand what this switch is aiming to do?

If the aim is to call local vs remote, I recommend breaking this up into maybe two unexported functions that clearly do what they do, with a condition here that decides based on peer id. That should also get rid of `interface {}` type as parameter into a concrete type.